eol.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH EOL 1 "September 19, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
eol \- program to translate text file end-of-line characters.
.SH SYNOPSIS
.B eol
.RI -c\  <conversion> \ [ \ infile \ [\ -o\  outfile \ ]\ ...\ ]
.br
.B tounix
.RI [\ -c\  <conversion> \ ]\ [\  infile \ [\ -o\  outfile \ ]\ ...\ ]
.br
.B tomac
.RI [\ -c\  <conversion> \ ]\ [\  infile \ [\ -o\  outfile \ ]\ ...\ ]
.br
.B todos
.RI [\ -c\  <conversion> \ ]\ [\  infile \ [\ -o\  outfile \ ]\ ...\ ]
.br
.B towin
.RI [\ -c\  <conversion> \ ]\ [\  infile \ [\ -o\  outfile \ ]\ ...\ ]
.SH DESCRIPTION
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBeol\fP translates end of line characters between different
operating system formats.
.B eol
will translate a list of files given on the command line. The
translations are by default done "in place" meaning the input file is the
same as the output file. This can be changed with the
.B \-o
command line option.
.PP
If no input or output files are specified then \fBeol\fP will translate
\fIstdin\fP to \fIstdout\fP.
.PP
\fIstdin\fP and \fIstdout\fP may be specfied explicitly by using \fB-\fP as
the input name and output name, respectively.
.PP
Unlike similiar programs, 
.B eol
doesn't care about what the current format of the text file
is. Therefore, translating a file twice will have no effect the second
time (ie, it will not corrupt the file).
.PP
.B tounix
is a shortcut for "eol -c tounix".
.br
.B tomac
is a shortcut for "eol -c tomac".
.br
.B todos
is a shortcut for "eol -c todos".
.br
.B towin
is a shortcut for "eol -c towin".
.SH OPTIONS
.TP
.BI \-c\  <conversion>
This commands eol (and the toXXX variants) to translate the files with the specified conversion format.
.PP
.RS
.I <conversion>
can be one of:
.TP
.B tounix
Translates to Unix style end of line characters (Linefeed)
.TP
.B tomac
Translates to Macintosh style end of line characters (Carriage-Return)
.TP
.B todos
Translates to DOS style end of line characters (Carriage-Return Linefeed)
.TP
.B towin
The same as
.B todos.
.RE

.TP
.BI \-o\  outfile
This makes the preceding input file be written to
.I outfile
instead of being translated in place.
.\" .SH SEE ALSO
.\" .BR bar (1),
.\" .BR baz (1).
.\" .br
.\" The programs are documented fully by
.\" .IR "The Rise and Fall of a Fooish Bar" ,
.\" available via the Info system.
.SH EXAMPLES
.TP
.B tounix macfile
translates
.B macfile
into unix text format.
.TP
.B tounix -c todos macfile
translates
.B macfile
into DOS text format.
.TP
.B eol -c tomac file1 file2 file3 -o file4 file5
translates
.B file1, file2, 
and
.B file5
into Macintosh text format.
.B file3
is also translated but the new translated version is saved into
.B file4
instead of being translated in place.

.SH BUGS
None, you pessimist!
.SH AUTHOR
This manual page was written by David Caldwell
 <david@porkrind.org>


ChangeLog
2024-10-06  David Caldwell  <david@porkrind.org>

	* Update to compile with modern gcc

	* Stop using some deprecated libc functions

	* Added -v and -h options (for version and help)

	* No arguments now means pipe through stdin/stdout

	* Make explicit stdin ("eol -") work

2001-09-26  David Caldwell  <david@indigita.com>

	* eol.cc (Eol): Removed debug statement I had left in by accident.

2001-09-20  David Caldwell  <david@indigita.com>

	* eol.1: Created manpage

	* makefile: Added manpage and links to install

	* version.h: Upped version to 1.0.1


eol.c
// $Header$
// -------=====================<<<< COPYRIGHT >>>>========================-------
//          Copyright (c) 1995-2024 David Caldwell,  All Rights Reserved.
//  See full text of copyright notice and limitations of use in file COPYRIGHT.h
// -------================================================================-------

#include "COPYRIGHT.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#include <libgen.h>
#include <limits.h>
#include <err.h>
#include "version.h"

#define CR '\r'
#define LF '\n'
#define qCR "\r"
#define qLF "\n"

struct {
    char *operation;
    char *name;
    char *eol;
} gEOL[] = {
    { "tomac",  "Macintosh (CR)", qCR     },
    { "tounix", "Unix (LF)",      qLF     },
    { "todos",  "DOS (CRLF)",     qCR qLF },
    { "towin",  "DOS (CRLF)",     qCR qLF },
    { NULL }
};

int Eol(char *inFileName,char *outFileName,char *eol);
int ConvertEOFs(char *in, char *out,int length, char *eol);
int move(char *oldpath, char *newpath,char *eol);

void Usage(char *program)
{
    printf("%s version %s\n",program,VersionString);
    printf("usage:\n"
           "   %s [-c <conversion>] [infile [-o outfile] ...]\n"
           "      For each infile, if '-o' is specified, 'infile' will be copy converted to 'outfile'.\n"
           "      Otherwise, 'infile' will be converted in-line.\n"
           "   %s -h   # Print help\n"
           "   %s -v   # Print version information\n"
           "\n"
           " <conversion> can be one of:\n",program,program,program);
    for (int i=0;gEOL[i].operation;i++)
        printf("    %-6s  -- convert End-Of-Lines to %s style.\n",gEOL[i].operation,gEOL[i].name);
    exit(1);
}

int streq(char *s,char *s1)
{
    return strcmp(s,s1) == 0;
}

FILE *OpenAndErr(char *name, char *mode)
{
    //special case for "-" filename.
    if (streq(name,"-")) {
        if (*mode == 'r')
            return stdin;
        if (*mode == 'w')
            return stdout;
    }

    FILE *file = fopen(name,mode);
    if (!file) {
        fprintf(stderr,"couldn't open '%s'.\n",name);
        exit(1);
    }
    return file;
}
int CheckErrors(FILE *f, char *name)
{
    if (ferror(f)) {
        fprintf(stderr,"Error using file '%s'.\n",name);
        return 1;
    }
    return 0;
}

int main (int c, char **v)
{
    char *program,*operation,*base;
    program = v[0];   // this doesn't change -- its what they said to run

    if((base=strrchr(v[0],'/')) || (base=strrchr(v[0],'\\')))
        program = base + 1;

#if defined(__WIN32__) || defined(_WIN32)
    for(char *cc=program; *cc; cc++)
        *cc = tolower(*cc);
    char *dot = strstr(program,".exe");
    if(dot)
        *dot = '\0';
#endif

    operation = program;

    int ch;
    while ((ch = getopt(c, v, "c:vh?")) != -1) {
        switch(ch) {
            case 'c': operation = optarg; break;
            case 'h':
            case '?': Usage(program);
            case 'v': printf("%s %s\n",program,VersionString); exit(0);

        }
    }
    c -= optind;
    v += optind;

    if (c==0) {
        static char *stdinout_args[] = { "-", "-o", "-" };
        v = stdinout_args;
        c = 3;
    }

    int style;
    for (style=0;gEOL[style].operation;style++)
        if (strcmp(gEOL[style].operation,operation) == 0)
            goto goodOperation;
    fprintf(stderr,"%s: Bad operation %s\n", program, operation);
    Usage(program);

  goodOperation:
    while(c) {
        char *inFileName = *v;
        v++;c--;
        char *outFileName = NULL;
        if (c) {
            if (strcmp(*v,"-o")==0) {
                v++;c--;
                if (!c) {
                    fprintf(stderr,"%s: -o requires a filename\n",program);
                    Usage(program);
                }
                outFileName = *v;
                v++;c--;
            }
        }

        Eol(inFileName,outFileName,gEOL[style].eol);
    }
}

int Eol(char *inFileName,char *outFileName,char *eol)
{
    FILE *in = OpenAndErr(inFileName,"rb");

    char tmpName[PATH_MAX];
    int inPlace = 0;
    FILE *out;
    if (!outFileName && streq(inFileName, "-"))
        outFileName = "-";
    if (!outFileName) {
        if (snprintf(tmpName, sizeof(tmpName), "%s-XXXXXX", basename(inFileName)) >= sizeof(tmpName))
            strcpy("eol-XXXXXX", tmpName);
        int ofd = mkstemp(tmpName);
        if (ofd < 0)
            err(EXIT_FAILURE, "Error creating temp file");
        out = fdopen(ofd, "w");
        outFileName = tmpName;
        inPlace = 1;
    } else
        out = OpenAndErr(outFileName,"wb");

#define BUFFER_SIZE 16384
    do {
        char inBuffer[BUFFER_SIZE];
        char outBuffer[BUFFER_SIZE*2]; // worst case unix->dos, with whole file of LFs.
        int read  = fread(inBuffer,1,BUFFER_SIZE,in);
        int write = ConvertEOFs(inBuffer,outBuffer,read,eol);
        fwrite(outBuffer,1,write,out);
        if (CheckErrors(in,inFileName) || CheckErrors(out,outFileName))
            exit(1);
    } while (!feof(in));
    fclose(out);
    fclose(in);

    if (inPlace) {
        if (!remove(inFileName)) {
            if (move(outFileName,inFileName,eol)) {
                fprintf(stderr,"Couldn't move '%s' to '%s'\n",outFileName, inFileName);
                return -1;
            }
        } else {
            fprintf(stderr,"Couldn't delete file '%s'.\n",inFileName);
            return -1;
        }
    }
    return 0;
}

int ConvertEOFs(char *in, char *out,int length, char *eol)
{
    static char last = 0;
    int outLength=0;
    int eolLength = strlen(eol);

    while (length--) {
        char current = *in++;
        if ((last == CR || last == LF) && (current == CR || current == LF) && last != current)
            current = 0; // dont count this char in next round--throw away CRLF pairs (or LFCR)
        else if (current == CR || current == LF) {
            memcpy(out,eol,eolLength);
            out += eolLength;
            outLength += eolLength;
        } else {
            *out++ = current;
            outLength++;
        }
        last = current;
    }
    return outLength;
}

#include <errno.h>

int move(char *oldpath, char *newpath,char *eol)
{
    int err=0;
    if ((err = rename(oldpath,newpath)) && errno == EXDEV) {
        // Cant move accross filesystems, so do it manually
        if (Eol(oldpath,newpath,eol)) {
            fprintf(stderr,"Couldn't copy file '%s' to '%s'\n",oldpath,newpath);
            return -1;
        }
        if (remove(oldpath)) {
            fprintf(stderr,"Couldn't delete '%s'.\n",oldpath);
            return -1;
        }
    } else if (err) {
        fprintf(stderr,"Couldn't rename '%s' to '%s'.\n",oldpath,newpath);
        return -1;
    }
    return 0;
}
